mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala

```diff
@@ -156,54 +153,22 @@ final class OneVsRestModel private[ml] (
     // Check schema
     transformSchema(dataset.schema, logging = true)
 
-    // determine the input columns: these need to be passed through
-    val origCols = dataset.schema.map(f => col(f.name))
-
-    // add an accumulator column to store predictions of all the models
-    val accColName = "mbc$acc" + UUID.randomUUID().toString
-    val initUDF = udf { () => Map[Int, Double]() }
-    val newDataset = dataset.withColumn(accColName, initUDF())
-
-    // persist if underlying dataset is not persistent.
-    val handlePersistence = dataset.storageLevel == StorageLevel.NONE
-    if (handlePersistence) {
-      newDataset.persist(StorageLevel.MEMORY_AND_DISK)
-    }
-
-    // update the accumulator column with the result of prediction of models
-    val aggregatedDataset = models.zipWithIndex.foldLeft[DataFrame](newDataset) {
-      case (df, (model, index)) =>
-        val rawPredictionCol = model.getRawPredictionCol
-        val columns = origCols ++ List(col(rawPredictionCol), col(accColName))
-
-        // add temporary column to store intermediate scores and update
-        val tmpColName = "mbc$tmp" + UUID.randomUUID().toString
-        val updateUDF = udf { (predictions: Map[Int, Double], prediction: Vector) =>
-          predictions + ((index, prediction(1)))
+    val predictUDF = udf { (features: Any) =>
+      var i = 0
+      var maxIndex = Double.NaN
+      var maxPred = Double.MinValue
+      while (i < models.length) {
+        val pred = models(i).predictRawAsFeaturesType(features)(1)
```

Comment:
Here serialize & broadcast models, is it safe and will it bring some overhead ?

Reply:
I thought on this, when a model is a distributed model, e.g, this model including a `Dataframe` or `RDD` data, then it cannot be broadcasted. In `OneVsRest`, it should support any kind of classifier, so we cannot suppose that the corresponding model of specified classifier can be broadcast.

Reply:
@WeichenXu123  Agree that we should take some distributed classification model like `DistributedLDAModel` in the future into account, but for now all classification models are not distributed.
What about let models tell us whether it is distributed (like adding a method named `isDistributed`), and determine the way of prediction in OVR?